### PR TITLE
feat: set TemplateRef in MUR spec

### PR DESCRIPTION
**Pull request by xcoulon**
Set the `TemplateRef` field in the namespaces
and cluster resources of the MUR during provisioning.

Updates https://issues.redhat.com/browse/CRT-500

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

**Comment by xcoulon** [comment]
E2E tests: https://github.com/codeready-toolchain/toolchain-e2e/pull/109

File: pkg/controller/usersignup/usersignup_controller.go [hunk 1]
```diff
@@ -425,14 +426,16 @@ func (r *ReconcileUserSignup) provisionMasterUserRecord(userSignup *toolchainv1a
 	namespaces := make([]toolchainv1alpha1.NSTemplateSetNamespace, len(nstemplateTier.Spec.Namespaces))
 	for i, ns := range nstemplateTier.Spec.Namespaces {
 		namespaces[i] = toolchainv1alpha1.NSTemplateSetNamespace{
-			Type:     ns.Type,
-			Revision: ns.Revision,
+			Type:        ns.Type,
+			Revision:    ns.Revision,
+			TemplateRef: nstemplatetiers.NewTierTemplateName(nstemplateTier.Name, ns.Type, ns.Revision),
```
Comment (MatousJobanek) [inline pkg/controller/usersignup/usersignup_controller.go:431]: We should rather use the TemplateRef from the tier than construct the name:
```suggestion
			TemplateRef: ns.TemplateRef,
```

File: pkg/controller/usersignup/usersignup_controller.go [hunk 2]
```diff
@@ -425,14 +426,16 @@ func (r *ReconcileUserSignup) provisionMasterUserRecord(userSignup *toolchainv1a
...
 		}
 	}
 	var clusterResources *toolchainv1alpha1.NSTemplateSetClusterResources
 	if nstemplateTier.Spec.ClusterResources != nil {
 		clusterResources = &toolchainv1alpha1.NSTemplateSetClusterResources{
-			Revision: nstemplateTier.Spec.ClusterResources.Revision,
+			Revision:    nstemplateTier.Spec.ClusterResources.Revision,
+			TemplateRef: nstemplatetiers.NewTierTemplateName(nstemplateTier.Name, nstemplatetiers.ClusterResources, nstemplateTier.Spec.ClusterResources.Revision),
```
Comment (MatousJobanek) [inline pkg/controller/usersignup/usersignup_controller.go:438]: same here:
```suggestion
			TemplateRef: nstemplateTier.Spec.ClusterResources.TemplateRef,
```

**Comment by xcoulon** [comment]
> If I understand it correctly this PR is only for provisioning; changes in ChangeTierRequest controller will be in the next PR, correct? If yes, then I'm fine with it and agree with multiple smaller PRs - I just wanted to clarify it 😉

yes, that's right: since you mentioned 3 steps in the Jira issue, I decided to address them separately. 

> 
> Could you please adjust also this test? https://github.com/codeready-toolchain/host-operator/blob/master/pkg/controller/usersignup/usersignup_controller_test.go#L113-L127
> and also the NSTemplateTier that is created as part of the test setup (there are missing the `TemplateRef` fields) https://github.com/codeready-toolchain/host-operator/blob/master/pkg/controller/usersignup/usersignup_controller_test.go#L54-L85

ok, let me do that.

**Comment by xcoulon** [comment]
> Could you please adjust also this test? https://github.com/codeready-toolchain/host-operator/blob/master/pkg/controller/usersignup/usersignup_controller_test.go#L113-L127
 > and also the NSTemplateTier that is created as part of the test setup (there are missing the `TemplateRef` fields) https://github.com/codeready-toolchain/host-operator/blob/master/pkg/controller/usersignup/usersignup_controller_test.go#L54-L85
> 

done in 34f6001

File: pkg/controller/usersignup/usersignup_controller.go [hunk 3]
```diff
@@ -13,6 +13,7 @@ import (
 
 	toolchainv1alpha1 "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
 	"github.com/codeready-toolchain/host-operator/pkg/configuration"
+	"github.com/codeready-toolchain/host-operator/pkg/templates/nstemplatetiers"
```
Comment (MatousJobanek) [inline pkg/controller/usersignup/usersignup_controller.go:16]: don't forget to remove the unused import
```suggestion
```

**Comment by MatousJobanek** [comment]
Thanks, @xcoulon, but you forgot to add it also for cluster resources: https://github.com/codeready-toolchain/host-operator/pull/201/files#diff-d7d6486f50f2c1cd16101f429e4ef843R80-R88

**Comment by xcoulon** [comment]
> Thanks, @xcoulon, but you forgot to add it also for cluster resources: https://github.com/codeready-toolchain/host-operator/pull/201/files#diff-d7d6486f50f2c1cd16101f429e4ef843R80-R88

done in 52efa2c

File: pkg/controller/changetierrequest/changetierrequest_controller.go [hunk 4]
```diff
@@ -144,14 +145,16 @@ func (r *ReconcileChangeTierRequest) changeTier(log logr.Logger, changeTierReque
 	namespaces := make([]toolchainv1alpha1.NSTemplateSetNamespace, len(nsTemplateTier.Spec.Namespaces))
 	for i, ns := range nsTemplateTier.Spec.Namespaces {
 		namespaces[i] = toolchainv1alpha1.NSTemplateSetNamespace{
-			Type:     ns.Type,
-			Revision: ns.Revision,
+			Type:        ns.Type,
+			Revision:    ns.Revision,
+			TemplateRef: nstemplatetiers.NewTierTemplateName(changeTierRequest.Spec.TierName, ns.Type, ns.Revision), // link to the TierTemplate resource, whose name is: `<tierName>-<nsType>-<revision>`,
```
Comment (MatousJobanek) [inline pkg/controller/changetierrequest/changetierrequest_controller.go:150]: The same thing again:
```suggestion
			TemplateRef: ns.TemplateRef,
```

File: pkg/controller/changetierrequest/changetierrequest_controller.go [hunk 5]
```diff
@@ -144,14 +145,16 @@ func (r *ReconcileChangeTierRequest) changeTier(log logr.Logger, changeTierReque
...
 		}
 	}
 	var clusterResources *toolchainv1alpha1.NSTemplateSetClusterResources
 	if nsTemplateTier.Spec.ClusterResources != nil {
 		clusterResources = &toolchainv1alpha1.NSTemplateSetClusterResources{
 			Revision: nsTemplateTier.Spec.ClusterResources.Revision,
+			//TemplateRef: nstemplatetiers.NewTierTemplateName(changeTierRequest.Spec.TierName, nstemplatetiers.ClusterResources, nsTemplateTier.Spec.ClusterResources.Revision), // link to the TierTemplate resource, whose name is: `<tierName>-<nsType>-<revision>`,
```
Comment (MatousJobanek) [inline pkg/controller/changetierrequest/changetierrequest_controller.go:157]: you need to set the TemplateRef here:
```suggestion
			TemplateRef: nsTemplateTier.Spec.ClusterResources.TemplateRef,
```

File: go.mod [hunk 6]
```diff
@@ -7,9 +7,10 @@ require (
 	github.com/codeready-toolchain/toolchain-common v0.0.0-20200504150510-6252228f3527
```
Comment (MatousJobanek) [inline go.mod:7]: you should update the toolchain-common to verify the TemplateRefs in the MUR spec

File: pkg/controller/changetierrequest/changetierrequest_controller_test.go [hunk 7]
```diff
@@ -274,15 +273,17 @@ func NewNSTemplateTier(tierName, revision, clusterResourcesRevision string, nsTy
...
+			Template:    templatev1.Template{},
+			TemplateRef: nstemplatetiers.NewTierTemplateName(tierName, nsType, revision),
 		}
 	}
 	var clusterResources *v1alpha1.NSTemplateTierClusterResources
 	if clusterResourcesRevision != "" {
 		clusterResources = &v1alpha1.NSTemplateTierClusterResources{
-			Revision: clusterResourcesRevision,
+			Revision:    clusterResourcesRevision,
+			TemplateRef: nstemplatetiers.NewTierTemplateName(tierName, nstemplatetiers.ClusterResources, revision),
```
Comment (MatousJobanek) [inline pkg/controller/changetierrequest/changetierrequest_controller_test.go:286]: you should use clusterResourcesRevision here
```suggestion
			TemplateRef: nstemplatetiers.NewTierTemplateName(tierName, nstemplatetiers.ClusterResources, clusterResourcesRevision),
```

**Comment by xcoulon** [comment]
great, thanks for the multiple reviews @MatousJobanek and @alexeykazakov 🙌

**Pull request merged**